cdap-tms/src/main/java/co/cask/cdap/messaging/store/AbstractMessageTable.java

```diff
@@ -0,0 +1,202 @@
+
+  @Override
+  public CloseableIterator<Entry> fetch(TopicId topicId, long startTime, int limit,
+                                        @Nullable Transaction transaction) throws IOException {
+    byte[] topic = MessagingUtils.toRowKeyPrefix(topicId);
+    byte[] startRow = new byte[topic.length + Long.BYTES];
+    Bytes.putBytes(startRow, 0, topic, 0, topic.length);
+    Bytes.putLong(startRow, topic.length, startTime);
+    byte[] stopRow = Bytes.stopKeyForPrefix(topic);
+    final CloseableIterator<RawMessageTableEntry> scanner = read(startRow, stopRow);
+    return new MessageTableCloseableIterator(scanner, limit, transaction);
+  }
+
+  @Override
+  public CloseableIterator<Entry> fetch(TopicId topicId, MessageId messageId, boolean inclusive, final int limit,
+                                        @Nullable final Transaction transaction) throws IOException {
+    byte[] topic = MessagingUtils.toRowKeyPrefix(topicId);
+    byte[] startRow = new byte[topic.length + Long.BYTES + Short.BYTES];
+    Bytes.putBytes(startRow, 0, topic, 0, topic.length);
+    Bytes.putLong(startRow, topic.length, messageId.getPublishTimestamp());
+    Bytes.putShort(startRow, topic.length + Long.BYTES, messageId.getSequenceId());
+    if (!inclusive) {
+      startRow = Bytes.incrementBytes(startRow, 1);
+    }
+    byte[] stopRow = Bytes.stopKeyForPrefix(topic);
+    final CloseableIterator<RawMessageTableEntry> scanner = read(startRow, stopRow);
+    return new MessageTableCloseableIterator(scanner, limit, transaction);
+  }
+
+  private static class MessageTableCloseableIterator extends AbstractCloseableIterator<Entry> {
+    private final CloseableIterator<RawMessageTableEntry> scanner;
+    private final Transaction transaction;
+    private boolean closed = false;
+    private int maxLimit;
+
+    MessageTableCloseableIterator(CloseableIterator<RawMessageTableEntry> scanner, int limit,
+                                  @Nullable Transaction transaction) {
+      this.scanner = scanner;
+      this.transaction = transaction;
+      this.maxLimit = limit;
+    }
+
+    @Override
+    protected Entry computeNext() {
+      if (closed || (maxLimit <= 0)) {
+        return endOfData();
+      }
+
+      while (scanner.hasNext()) {
+        RawMessageTableEntry tableEntry = scanner.next();
+        Entry entry = new DefaultMessageTableEntry(tableEntry.getKey(), tableEntry.getPayload(), tableEntry.getTxPtr());
+        RESULT status = isVisible(entry, transaction);
+        if (status == RESULT.ACCEPT) {
+          maxLimit--;
+          return entry;
+        }
+
+        if (status == RESULT.HOLD) {
+          break;
+        }
+      }
+      return endOfData();
+    }
+
+    @Override
+    public void close() {
+      try {
+        scanner.close();
+      } finally {
+        endOfData();
+        closed = true;
+      }
+    }
+  }
+
+  @Override
+  public void store(Iterator<Entry> entries) throws IOException {
+    persist(new StoreIterator(entries));
```

Comment:
You can make `StoreIterator` mutable so that we don't need to create new instance of it. Just have a `StoreIterator reset(long lastWriteTimestamp, long writeTimestamp, Iterator<Entry> entries)` method will do.